### ISSUE TITLE
Validation error while posting the workflow

### ISSUE THREAD

**Issue by komatiraju032** [Netflix/conductor#2212]
I am trying to add the workflow using `/metadata/workflow` endpoint. When I make the request it's failing with validation errors.

I am using [kitchensink workflow](https://netflix.github.io/conductor/labs/kitchensink/#definition).

**Curl Request**
```
curl -X POST "http://localhost:8080/api/metadata/workflow" -H  "accept: */*" -H  "Content-Type: application/json" -d "{\"name\":\"kitchensink\",\"description\":\"kitchensink workflow\",\"version\":1,\"tasks\":[{\"name\":\"task_1\",\"taskReferenceName\":\"task_1\",\"inputParameters\":{\"mod\":\"\${workflow.input.mod}\",\"oddEven\":\"\${workflow.input.oddEven}\"},\"type\":\"SIMPLE\"},{\"name\":\"event_task\",\"taskReferenceName\":\"event_0\",\"inputParameters\":{\"mod\":\"\${workflow.input.mod}\",\"oddEven\":\"\${workflow.input.oddEven}\"},\"type\":\"EVENT\",\"sink\":\"conductor\"},{\"name\":\"dyntask\",\"taskReferenceName\":\"task_2\",\"inputParameters\":{\"taskToExecute\":\"\${workflow.input.task2Name}\"},\"type\":\"DYNAMIC\",\"dynamicTaskNameParam\":\"taskToExecute\"},{\"name\":\"oddEvenDecision\",\"taskReferenceName\":\"oddEvenDecision\",\"inputParameters\":{\"oddEven\":\"\${task_2.output.oddEven}\"},\"type\":\"DECISION\",\"caseValueParam\":\"oddEven\",\"decisionCases\":{\"0\":[{\"name\":\"task_4\",\"taskReferenceName\":\"task_4\",\"inputParameters\":{\"mod\":\"\${task_2.output.mod}\",\"oddEven\":\"\${task_2.output.oddEven}\"},\"type\":\"SIMPLE\"},{\"name\":\"dynamic_fanout\",\"taskReferenceName\":\"fanout1\",\"inputParameters\":{\"dynamicTasks\":\"\${task_4.output.dynamicTasks}\",\"input\":\"\${task_4.output.inputs}\"},\"type\":\"FORK_JOIN_DYNAMIC\",\"dynamicForkTasksParam\":\"dynamicTasks\",\"dynamicForkTasksInputParamName\":\"input\"},{\"name\":\"dynamic_join\",\"taskReferenceName\":\"join1\",\"type\":\"JOIN\"}],\"1\":[{\"name\":\"fork_join\",\"taskReferenceName\":\"forkx\",\"type\":\"FORK_JOIN\",\"forkTasks\":[[{\"name\":\"task_10\",\"taskReferenceName\":\"task_10\",\"type\":\"SIMPLE\"},{\"name\":\"sub_workflow_x\",\"taskReferenceName\":\"wf3\",\"inputParameters\":{\"mod\":\"\${task_1.output.mod}\",\"oddEven\":\"\${task_1.output.oddEven}\"},\"type\":\"SUB_WORKFLOW\",\"subWorkflowParam\":{\"name\":\"sub_flow_1\",\"version\":1}}],[{\"name\":\"task_11\",\"taskReferenceName\":\"task_11\",\"type\":\"SIMPLE\"},{\"name\":\"sub_workflow_x\",\"taskReferenceName\":\"wf4\",\"inputParameters\":{\"mod\":\"\${task_1.output.mod}\",\"oddEven\":\"\${task_1.output.oddEven}\"},\"type\":\"SUB_WORKFLOW\",\"subWorkflowParam\":{\"name\":\"sub_flow_1\",\"version\":1}}]]},{\"name\":\"join\",\"taskReferenceName\":\"join2\",\"type\":\"JOIN\",\"joinOn\":[\"wf3\",\"wf4\"]}]}},{\"name\":\"search_elasticsearch\",\"taskReferenceName\":\"get_es_1\",\"inputParameters\":{\"http_request\":{\"uri\":\"http://localhost:9200/conductor/_search?size=10\",\"method\":\"GET\"}},\"type\":\"HTTP\"},{\"name\":\"task_30\",\"taskReferenceName\":\"task_30\",\"inputParameters\":{\"statuses\":\"\${get_es_1.output..status}\",\"workflowIds\":\"\${get_es_1.output..workflowId}\"},\"type\":\"SIMPLE\"}],\"outputParameters\":{\"statues\":\"\${get_es_1.output..status}\",\"workflowIds\":\"\${get_es_1.output..workflowId}\"},\"ownerEmail\":\"example@email.com\",\"schemaVersion\":2}"
```
**Error:**
```
{
  "status": 400,
  "message": "Validation failed, check below errors for detail.",
  "retryable": false,
  "validationErrors": [
    {
      "path": "registerWorkflowDef.workflowDef.tasks[3].decisionCases[1].<map value>[0].forkTasks[1][0]",
      "message": "workflowTask: task_11 task definition is not defined"
    },
    {
      "path": "registerWorkflowDef.workflowDef.tasks[3].decisionCases[0][0]",
      "message": "workflowTask: task_4 task definition is not defined"
    },
    {
      "path": "registerWorkflowDef.workflowDef.tasks[3].decisionCases[1][0].forkTasks[0][0]",
      "message": "workflowTask: task_10 task definition is not defined"
    },
    {
      "path": "registerWorkflowDef.workflowDef.tasks[0]",
      "message": "workflowTask: task_1 task definition is not defined"
    },
    {
      "path": "registerWorkflowDef.workflowDef.tasks[3].decisionCases[1].<map value>[0].forkTasks[0].<list element>[0]",
      "message": "workflowTask: task_10 task definition is not defined"
    },
    {
      "path": "registerWorkflowDef.workflowDef.tasks[3].decisionCases[1].<map value>[0].forkTasks[1].<list element>[0]",
      "message": "workflowTask: task_11 task definition is not defined"
    },
    {
      "path": "registerWorkflowDef.workflowDef.tasks[3].decisionCases[1].<map value>[0].forkTasks[0][0]",
      "message": "workflowTask: task_10 task definition is not defined"
    },
    {
      "path": "registerWorkflowDef.workflowDef.tasks[5]",
      "message": "workflowTask: task_30 task definition is not defined"
    },
    {
      "path": "registerWorkflowDef.workflowDef.tasks[3].decisionCases[1][0].forkTasks[0].<list element>[0]",
      "message": "workflowTask: task_10 task definition is not defined"
    },
    {
      "path": "registerWorkflowDef.workflowDef.tasks[3].decisionCases[1][0].forkTasks[1].<list element>[0]",
      "message": "workflowTask: task_11 task definition is not defined"
    },
    {
      "path": "registerWorkflowDef.workflowDef.tasks[3].decisionCases[1][0].forkTasks[1][0]",
      "message": "workflowTask: task_11 task definition is not defined"
    },
    {
      "path": "registerWorkflowDef.workflowDef.tasks[3].decisionCases[0].<map value>[0]",
      "message": "workflowTask: task_4 task definition is not defined"
    }
  ]
}
```

**Comment by kishorebanala** [comment]
Hey @komatiraju032, Looking at the validation messages above, I believe the tasks are not registered before registering a workflow. The validation order is to ensure that the tasks are defined before being used in a workflow.

**Comment by komatiraju032** [comment]
@kishorebanala But the same request working in conductor 2.x.

**Comment by devinderpal** [comment]
I am also facing the same issue. In my case, tasks have been registered already. The code was working fine with 2.x but not working with v3.0.2

**Comment by devinderpal** [comment]
After some troubleshooting, it's failing at the below code in WorkflowTaskValidConstraint.
Even though I am using "mysql" persistence, ValidationContext.getMetadataDAO() = com.netflix.conductor.redis.dao.**RedisMetadataDAO**@...

Why is it pointing to Redis? It should be MySQLMetadataDAO

```
if (ValidationContext.getMetadataDAO().getTaskDef(workflowTask.getName()) == null) {
    //check if task type is ephemeral
    TaskDef task = workflowTask.getTaskDefinition();
    if (task == null) {
        valid = false;
        String message = String
            .format("workflowTask: %s task definition is not defined", workflowTask.getName());
        context.buildConstraintViolationWithTemplate(message).addConstraintViolation();
    }
}
```

**Comment by devinderpal** [comment]
After some more investigation, looks like the root cause is that `conductor-mysql-persistence-*.jar` is not in the conductor-server-*-boot.jar. So it can't load MySQLMetadataDAO. Even postgres jar is not there. So currently v3.x wouldn't work correctly with these databases.

**Comment by apanicker-nflx** [comment]
This should be fixed as part of #2222
